docs/proposals-accepted/202304-cardinalitymanagment.md

```diff
@@ -0,0 +1,132 @@
+   
+    * Add new data points (series, labels, and label-value pairs).
+    * Compute the statistics required by the API.
+    * Serialize the data into a JSON format according to the API design.
+
+3. We will define a new HTTP endpoint in the Thanos receiver to expose the API,as `/api/v1/cardinality`.
+
+4. We will Implement a new API handler function that:
+   
+    * Accepts optional query parameters, such as top , for filtering and sorting the results.
+    * Calls the data structure's methods to compute the required statistics.
+    * Serializes the results into JSON format and returns it as an HTTP response.
+    * Handles potential partial results by setting the isPartial field in the response and providing an appropriate error message.
+
+5. (Optional)We can add a flag to the Thanos receiver, allowing users to enable or disable the API.
+
+### Alternatives
+* Alternative approach could be through Prometheus's [`/api/v1/series`](https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers) but, it not useful for large scale or large number of labels as when using the `/api/v1/series` endpoint, the Prometheus server must iterate over all the time series data to identify the series that match the label selectors in the match[] parameter. This can be a computationally intensive operation.
+* We can also perform  another API design, where we will have two seperate endpoints for label names cardinality and label values cardinality.Let us take a look at it:
+
+   * Label Names endpoint: This endpoint will be responsible for returning the realtime label names cardinality . It will count distinct label values per label name.<br>
+  Response: <br>
+  ```
+  {
+  "label_values_count_total": <number>,
+  "label_names_count": <number>,
+  "cardinality": [
+    {
+      "label_name": <string>,
+      "label_values_count": <number>
+    }
+  ]
+  }
+  ```
+
+
+  * Label values cardinality endpoint: This endpoint will return the label values cardinality associated with request parameter `label_names[]`. It will return the series count per label value for each label in the request parameter `label_names[]`.<br>
+
+  ```{
+  "series_count_total": <number>,
+  "labels": [
+    {
+      "label_name": <string>,
+      "label_values_count": <number>,
+      "series_count": <number>,
+      "cardinality": [
+        {
+          "label_value": <string>,
+          "series_count": <number>
+        }
+      ]
+    }
+  ]
+  } 
+  ```
+  
+Now, both these endpoints can have a `limit` parameter which will specify number of items in the response.
+
+#### Pros
+1. It provides more granularity about series data.Also, we can get specific information that we need.Users can choose to query label names, label values, or both depending on their requirements.
+2. Users can provide optional PromQL selectors to filter series to analyze, allowing them to focus on specific metrics or series of interest.
```

Comment:
Why our prefered approach cannot take PromQL selectors? It should work for both APIs, right?